test-case: check aplay/arecord before and after runtime pm

fredoh9 opened the pull request:
Before and after runtime pm state transition, check aplay or arecord is
working. The delay is chose from range of 3 to 15 seconds. When the delay
is long enough to have 'suspended' state, then check for the state as well.

Add capture pipeline support for runtime pm test

aplay/arecord can be used by setting -m option. Add scanning dmesg before
switching pipeline.


For reviewers, initially I wanted to merge this two test cases with new input parameter, but it bacame too ugly to read. One more loop and multiple checking mode etc.

xiulipan commented:
@fredoh9 What is the difference with the `check-runtime-pm-status.sh`. They both try to check if the FW can work normally when runtime pm happen.

fredoh9 commented:
@xiulipan Per pipeline,
`check-runtime-pm-status.sh`
aplay -> atop stop -> sleep for enough time -> check runtime pm status

`check-runtime-pm-status-boundary.sh`
aplay(or arecord) -> atop stop -> sleep range of [3 5 7 10 15] -> check runtime pm status if resired time is passed -> check aplay(or arecord) 

New script will especially check if FW can work at boundary of runtime pm.

stripes416 commented on test-case/check-runtime-pm-boundary.sh:
```diff
@@ -0,0 +1,165 @@
+## Case step:
+##    1. start aplay/arecord
+##    2. stop aplay/arecord
+##    3. sleep of [3 5 7 10 15]
```
This is a sleep period in seconds right?

Bin-QA commented on test-case/check-runtime-pm-boundary.sh:
```diff
@@ -0,0 +1,165 @@
+                [[ $? -ne 0 ]] && dloge "$cmd process for pcm $pcm is not alive" && exit 1
+
+                kill -9 $pid && wait $pid 2>/dev/null
+                dlogi "$cmd end"
```
I don't think this operation is different with case: [check-runtime-pm-status](https://github.com/thesofproject/sof-test/blob/master/test-case/check-runtime-pm-status.sh)  

you can see when you run this logic in the loop, so next round will run the cmd again  
So I will suggest you add the **loop** option for the original case instead of this change  
if `$i` try to run and trigger the error at here, I think original case will trigger at `$i + 1` at begin  
when add **loop** option

Bin-QA commented on test-case/check-runtime-pm-boundary.sh:
```diff
@@ -0,0 +1,165 @@
+    if [[ $result != suspended ]]; then
+        dloge "$cmd process for pcm $pcm runtime status is not suspended as expected"
+        exit 1
+    fi
```
the original case main logic for the loop
```
loop begin
$cmd # to run cmd
check $runtime_status
is active
    stop $cmd
    check $runtime_status
    is suspended
loop end
```
your new case logic:
```
loop begin
$cmd # to run cmd
check $runtime_status
is active
    stop $cmd
    check $runtime_status
    is suspended
    $cmd # to run cmd
loop end
is suspended
```

So I couldn't see the different between the original 
Because those check will cover by the loop
so if you want check `suspended` you can add at original loop begin  

The different between this case with original, I think this case will catch at `$i`, the original will catch at `$i + 1`

fredoh9 replied:
@Bin-QA the main purpose of this test is check aplay/arecord again just before/after runtime pm, having said that probably sleep [3 5 7] are most important delay time to start playback or capture again. There was known issue, it didn't work before.

Adding this logic to known runtime pm test case, it looks to me messy.

Bin-QA replied:
how about give the random time range and increase the loop

xiulipan commented:
@fredoh9 I agreed with @Bin-QA I did not see any difference of code logic run with https://github.com/thesofproject/sof-test/blob/master/test-case/check-runtime-pm-status.sh
The PM status check also include check DSP is active when pipeline run and check DSP is suspend when pipeline idle for sometime. What will be difference when you wait for [3 5 7] here?

keqiaozhang commented:
@fredoh9 
This is a corner case, when sof is idle, runtime_status will be changed like this: active->suspending->suspended, so I think "suspending" status is the boundary you want to test.
You can test it like this:
```
while true
do
status=`tail $runtime_status`
echo $status
if [ $status == 'suspending' ]
then
        run aplay/arecord
        check error
fi
done
```

fredoh9 commented:
@keqiaozhang I have a plan to re-visit this test case today. Will update again and remove WIP.

fredoh9 commented:
@xiulipan @keqiaozhang @Bin-QA @libinyang 
I rebased to latest master and updated this again. this is highlighted changes.

1. Need more frequent check of the variable. Updated polling time to 100ms
2. When runtime pm status get suspended, trigger playback/capture again with duration of 1 second
4. commit text changed, file name also is changed

Bin-QA commented on test-case/check-runtime-pm-double-active.sh:
```diff
@@ -0,0 +1,145 @@
+func_opt_parse_option $*
+tplg=${OPT_VALUE_lst['t']}
+loop_count=${OPT_VALUE_lst['l']}
+[[ -z $tplg ]] && dloge "Miss tplg file to run" && exit 1
```
exit 2

Bin-QA commented on test-case/check-runtime-pm-double-active.sh:
```diff
@@ -0,0 +1,145 @@
+
+    cmd="${APP_LST[$type]}"
+    dummy_file="${DEV_LST[$type]}"
+    [[ -z $cmd ]] && dloge "$type is not supported, $cmd, $dummy_file" && exit 1
```
exit 2

Bin-QA commented on test-case/check-runtime-pm-double-active.sh:
```diff
@@ -0,0 +1,145 @@
+        fi
+    done
+
+    sof-kernel-log-check.sh 0
```
with `sof-kernel-log-check.sh 0` you also need `sudo dmesg -C` before loop

Bin-QA commented on test-case/check-runtime-pm-double-active.sh:
```diff
@@ -0,0 +1,145 @@
+            exit 1
+        fi
+    done
+    dlogi "dsp suspended in $i try, ${i}00 msec"
```
how about use `return` instead of `break`
then after loop, you can do the error process

libinyang replied:
@Bin-QA Do you mean `exit` or `break`

fredoh9 replied:
@Bin-QA once DSP status is suspended, `break ` the loop and measure the time elapsed is the point. Error handling is pretty much timeout only.

libinyang commented on test-case/check-runtime-pm-double-active.sh:
```diff
@@ -0,0 +1,145 @@
+            kill -9 $pid && wait $pid 2>/dev/null
+            dlogi "$cmd killed"
+            func_check_dsp_status ${OPT_VALUE_lst['d']}
+            result=`sof-dump-status.py --dsp_status 0`
```
`func_check_dsp_status` has already make sure status is "suspend", otherwise it will exit. So the below `if` will never happen.

fredoh9 commented on test-case/check-runtime-pm-double-active.sh:
```diff
@@ -0,0 +1,145 @@
+        # we have a third-party pci sound card installed.
+        [[ $(sof-dump-status.py --dsp_status 0) == "suspended" ]] && break
+        sleep 0.1
+        if [ $i -eq $iloop ]; then
```
this is good point!

libinyang commented:
LGTM
